plugins/webp-uploads/tests/test-picture-element.php

```diff
@@ -115,32 +111,68 @@ public function test_maybe_wrap_images_in_picture_element( bool $jpeg_and_webp,
+				'picture_element' => true,
+				'expected_html'   => '<img width="{{img-width}}" height="{{img-height}}" src="{{img-src}}" class="wp-image-{{img-attachment-id}}" alt="{{img-alt}}" decoding="async" loading="lazy" srcset="{{img-srcset}}" sizes="{{img-sizes}}" />',
 			),
 			'only jpeg enabled'        => array(
-				'jpeg_and_webp'          => true,
-				'picture_element'        => false,
-				'expect_picture_element' => false,
-				'expected_html'          => '<img width="{{img-width}}" height="{{img-height}}" src="{{img-src}}" class="wp-image-{{img-attachment-id}}" alt="{{img-alt}}" decoding="async" loading="lazy" srcset="{{img-srcset}}" sizes="{{img-sizes}}" />',
+				'fallback_jpeg'   => true,
+				'picture_element' => false,
+				'expected_html'   => '<img width="{{img-width}}" height="{{img-height}}" src="{{img-src}}" class="wp-image-{{img-attachment-id}}" alt="{{img-alt}}" decoding="async" loading="lazy" srcset="{{img-srcset}}" sizes="{{img-sizes}}" />',
 			),
 			'neither enabled'          => array(
-				'jpeg_and_webp'          => false,
-				'picture_element'        => false,
-				'expect_picture_element' => false,
-				'expected_html'          => '<img width="{{img-width}}" height="{{img-height}}" src="{{img-src}}" class="wp-image-{{img-attachment-id}}" alt="{{img-alt}}" decoding="async" loading="lazy" srcset="{{img-srcset}}" sizes="{{img-sizes}}" />',
+				'fallback_jpeg'   => false,
+				'picture_element' => false,
+				'expected_html'   => '<img width="{{img-width}}" height="{{img-height}}" src="{{img-src}}" class="wp-image-{{img-attachment-id}}" alt="{{img-alt}}" decoding="async" loading="lazy" srcset="{{img-srcset}}" sizes="{{img-sizes}}" />',
 			),
 		);
 	}
 
+	public function test_picture_source_only_have_additional_mime_not_jpeg_and_return_jpeg_fallback(): void {
+		$mime_type = 'image/webp';
+		if ( ! wp_image_editor_supports( array( 'mime_type' => $mime_type ) ) ) {
+			$this->markTestSkipped( "Mime type $mime_type is not supported." );
+		}
+
+		// Create some content with the image.
+		$image = wp_get_attachment_image(
+			self::$image_id,
+			'large',
+			false,
+			array(
+				'class' => 'wp-image-' . self::$image_id,
+				'alt'   => 'Green Leaves',
+			)
+		);
+
+		$img_markup = apply_filters( 'the_content', $image );
+
+		$img_processor = new WP_HTML_Tag_Processor( $img_markup );
+		$this->assertTrue( $img_processor->next_tag( array( 'tag_name' => 'IMG' ) ), 'There should be an IMG tag.' );
+		$img_src    = $img_processor->get_attribute( 'src' );
+		$img_srcset = $img_processor->get_attribute( 'srcset' );
+
+		// Apply picture element support.
+		$this->opt_in_to_picture_element();
+
+		$picture_markup    = apply_filters( 'the_content', $image );
+		$picture_processor = new WP_HTML_Tag_Processor( $picture_markup );
+
+		$picture_processor->next_tag( array( 'tag_name' => 'IMG' ) );
+		$this->assertSame( $img_src, $picture_processor->get_attribute( 'src' ), 'Make sure the IMG and Picture IMG have same image src.' );
```

Comment:
Is this right? The `$img_src` should be the AVIF/WebP image. The `IMG` inside the `PICTURE` should rather be the JPEG.

Reply:
This should add assertions that the `src` ends with `.jpg` and the `srcset` contains `.jpg`.

Reply:
@westonruter The initial image is `JPEG` image so the `$img_src` return `.jpg` IMG src.

Reply:
Why it return the `JPEG` images in unit tests?

`add_filter( 'the_content', 'webp_uploads_update_image_references', 13 );` runs that replace `JPEG` to it's additional mime.

```php
// Bail early if request is not for the frontend.
if ( ! webp_uploads_in_frontend_body() ) {
	return $content;
}
```

This condition return early for the unit tests so it will return `JPEG` instead of `WEBP` images in unit test.

Reply:
Ok, so then this should be mocked so that `webp_uploads_in_frontend_body()` returns true.

Reply:
Mocked isn't the right word. I mean, the environment should be set up so that `webp_uploads_in_frontend_body()` returns true. It seems this can be forced with code like:

```php
$GLOBALS['wp_actions']['template_redirect'] = 1;
```

Reply:
I used `mock_frontend_body_hooks()` in https://github.com/WordPress/performance/pull/1408/commits/e0e30eff3faf1b0d7835f8b3d7be3babf4b8d3b2 it was introduce in plugin while before.

It will return the WebP now. 🎉